### PR TITLE
upgrade pyarrow from 12 to 14

### PR DESCRIPTION
**Pull request by severo**
We temporarily stop supporting array features until `datasets` fixes the issue on their side.

### CODE REVIEW

**Review by albertvillanova — CHANGES_REQUESTED** [review]
Thank you. I think you forgot to update it in some of the components: api, rows,...

**Comment by severo** [comment]
> I think you forgot to update it in some of the components: api, rows,...

in fact, they were requiring pyarrow 11, which is why it was stuck to a lower version.

But... it seems like the upgrade will not be that simple: https://github.com/huggingface/datasets-server/actions/runs/6823050840/job/18556219132

Looking at it

**Review by albertvillanova — COMMENTED** [review]
> But... it seems like the upgrade will not be that simple

I see... I'm having a look at `datasets` side...

**Review by albertvillanova — COMMENTED** [review]
I see, the issue is in our `viewer_utils`...

File: libs/libcommon/pyproject.toml [hunk 1]
```diff
@@ -18,6 +18,7 @@ numba = "0.56.4"
 orjson = "^3.8.6"
 pandas = "~2.0.3"
 psutil = "^5.9.4"
+pyarrow = "^14.0.0"
```
[the remark this answers is not in the document]
Comment (severo) [reply libs/libcommon/pyproject.toml:21]: I'll fix it

**Comment by severo** [comment]
yes, to reproduce:

```
import numpy as np
from datasets import (Array2D, Dataset, Features)
feature_type = Array2D(shape=(2, 2), dtype="float32")
content = np.zeros((2, 2), dtype="float32")
features = Features({"col": feature_type})
dataset = Dataset.from_dict({"col": [content]}, features=features)
```

**Comment by severo** [comment]
Should I open an issue in datasets? Or am I misusing it? It's a bit too cryptic for me

File: libs/libcommon/pyproject.toml (hunk 1 above)
[the remark this answers is not in the document]
Comment (severo) [reply libs/libcommon/pyproject.toml]: done

**Review by albertvillanova — CHANGES_REQUESTED** [review]
> Should I open an issue in datasets?

Yes, I understand now: they have disabled their `PyExtensionType` and we use it in `datasets` for arrays...

**Comment by severo** [comment]
OK, I opened https://github.com/huggingface/datasets/issues/6396

**Comment by severo** [comment]
Once merged, we'll have to revert https://github.com/huggingface/datasets-server/pull/2090/commits/0f58abb96150e28f7dca12aa3a5d85e869535501 if #2090 is merged before.

**Review by lhoestq — APPROVED** [review]
Thanks for the reactivity !
I'm fine with disabling ArrayND types for now using 14.0.1

**Comment by severo** [comment]
>  I'm fine with disabling ArrayND types for now using 14.0.1

what does it mean? letting the error propagate, and remove it from the tests?

**Comment by lhoestq** [comment]
Let's comment the failing tests for now. I guess `pyarrow` will ship a fix that introduces safe extension arrays soon

**Comment by lhoestq** [comment]
(or maybe not, but in this case `datasets` should stop using `PyExtensionType`)

**Review by albertvillanova — COMMENTED** [review]
So we are no longer supporting datasets with arrays features...

**Comment by lhoestq** [comment]
Yes until `datasets` fixes it

**Comment by severo** [comment]
Exactly. I think that it's better to remove support for array features than to let the vulnerability.

**Review by albertvillanova — APPROVED** [review]

**Comment by severo** [comment]
Thanks for the comments and reviews!

**Pull request merged**